mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/MassIndexerImpl.java

```diff
@@ -0,0 +1,250 @@
+import org.hibernate.search.mapper.orm.cfg.impl.HibernateOrmConfigurationPropertySource;
+import org.hibernate.search.mapper.orm.massindexing.MassIndexer;
+import org.hibernate.search.mapper.orm.impl.HibernateSearchContextService;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.orm.mapping.spi.HibernateOrmMapping;
+import org.hibernate.search.mapper.orm.massindexing.monitor.MassIndexerProgressMonitor;
+import org.hibernate.search.mapper.orm.massindexing.monitor.impl.JMXRegistrar;
+import org.hibernate.search.mapper.orm.massindexing.monitor.impl.SimpleIndexingProgressMonitor;
+import org.hibernate.search.util.impl.common.Executors;
+import org.hibernate.search.util.impl.common.LoggerFactory;
+import org.hibernate.search.util.impl.common.StringHelper;
+
+/**
+ * Prepares and configures a BatchIndexingWorkspace to start rebuilding
+ * the indexes for all entity instances in the database.
+ * The type of these entities is either all indexed entities or a
+ * subset, always including all subtypes.
+ *
+ * @author Sanne Grinovero
+ */
+public class MassIndexerImpl implements MassIndexer {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private static final ConfigurationProperty<Boolean> JMX_ENABLED =
+			ConfigurationProperty.forKey( SearchOrmSettings.Radicals.JMX_ENABLED )
+					.asBoolean()
+					.withDefault( SearchOrmSettings.Defaults.JMX_ENABLED )
+					.build();
+
+	private final HibernateOrmMapping mapping;
+	private final SessionFactoryImplementor sessionFactory;
+	private final String tenantIdentifier;
+
+	protected final Set<Class<?>> rootEntities;
+
+	// default settings defined here:
+	private int typesToIndexInParallel = 1;
+	private int documentBuilderThreads = 6; //loading the main entity, also responsible for loading of lazy @IndexedEmbedded collections
+	private int objectLoadingBatchSize = 10;
+	private long objectsLimit = 0; //means no limit at all
+	private CacheMode cacheMode = CacheMode.IGNORE;
+	private boolean optimizeAtEnd = true;
+	private boolean purgeAtStart = true;
+	private boolean optimizeAfterPurge = true;
+	private MassIndexerProgressMonitor monitor;
+	private int idFetchSize = 100; //reasonable default as we only load IDs
+	private Integer idLoadingTransactionTimeout;
+
+
+
+	public MassIndexerImpl(SessionFactoryImplementor sessionFactory, String tenantIdentifier, Class<?>... entities) {
+		this.sessionFactory = sessionFactory;
+		this.tenantIdentifier = tenantIdentifier;
+
+		HibernateSearchContextService searchContextService = sessionFactory.getServiceRegistry().getService( HibernateSearchContextService.class );
+		this.mapping = searchContextService.getMapping();
+		this.rootEntities = toRootEntities( mapping, entities );
+
+		// TODO Maybe it is a little bit complex
+		ConfigurationService configurationService = sessionFactory.getServiceRegistry().getService( ConfigurationService.class );
```

Comment:
Actually I've not found a better way to read the property. I don't like the one I implemented here.

Reply:
As mentioned, JMX support was never part of the scope. Please remove this commit.

As to configuration properties, they should be read at startup, and stored somewhere. This kind of runtime configuration parsing should not be necessary at all.